README.md

```diff
@@ -7,10 +7,30 @@ The best way to use it is via [vaadin.com/start](https://vaadin.com/start) - you
 
 To access it directly from github, clone the repository and import the project to the IDE of your choice as a Maven project. You need to have Java 8 or 11 installed.
 
-Run using `mvn -Pprepare-osgi-container -Prun-osgi-container -Pproduction verify` and open [http://localhost:8080](http://localhost:8080) in browser.
-This command will run profiles which downloads for you Felix OSGi container and installs `felix-jetty` 
-as OSGi web container. It deploys all necessary bundles to Felix and start it synchronously.
-The container will be stopped  when you stop your maven build (e.g. by killing it via `Ctrl+C`).
+Main Command: `mvn clean install`
```

Comment:
This command just fails for me.

But may be it doesn't really matter since I'm going to create another PR

Reply:
it fails on an [integration-test](https://github.com/vaadin/base-starter-flow-osgi/blob/5f78e2fe208a36816369e51ea6f0386190a080fe/starter.flow.better/src/test/java/com/example/starter/flow/better/Test.java) 

you may have changed to much. i am absolutely sure that this builds and tests work with older version

Reply:
`Lookup.lookup(ResourceProvider.class)` has to return not null object otherwise OSGi integration is simply totally broken.

Even though `Lookup.lookup` may be changed to `Optional`  but this is a different question and most likely it will complicate the usage.

Reply:
Our IT tests work fine : `ResourceProvider` instance is available in lookup.

I had similar issue in our IT tests but I was not able to figure out what's going on there.
Apparently it's related to the fact _HOW_ the servlet is registered.
In case of bundle Activator : https://github.com/vaadin/flow/blob/2.4/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/Activator.java , 
https://github.com/vaadin/flow/blob/2.4/flow-tests/test-root-context/bnd.bnd#L9
then service https://github.com/vaadin/flow/blob/feature/osgi/flow-osgi/src/main/java/com/vaadin/flow/osgi/support/OSGiResourceProvider.java#L41 is not yet available at the servlet registration point.

But if I register a servlet this way: https://github.com/vaadin/flow/blob/feature/osgi/flow-tests/test-root-context/src/main/java/com/vaadin/flow/osgi/Activator.java#L48
then everything works fine.

So it's some race condition which is most likely also the case for you as well.

May be you know how to make this work properly ?
May be resource provider should be registered via a Bundle activator and not via Declarative service.

This 
```
@Reference
	ResourceProvider pr;
```
is more a workaround and not really a solution: you should not know anything about `ResourceProvider` which is internal SPI.

Reply:
i know that this is just an workaround! but it helps to show what the issue is.

This is an often as question: [best awnser](https://stackoverflow.com/questions/32455624/order-of-deployment-of-bundles)

You simply can not expect that a bundle/ service is exists. If you uninstall a bundle A1 with a service S1 and install a new version A2  at runtime all services of bundle B that needs this S1 have must be deactivated on uninstall and activated again on install.

DS will do all this for you.

Reply:
Well, it relates to the order and it doesn't.

I don't rely on any order when I ask for `ResourceProvider`. 
The service is asked at runtime directly : https://github.com/vaadin/flow/blob/feature/osgi/flow-server/src/main/java/com/vaadin/flow/server/osgi/OSGiAccess.java#L95 on demand.

So there is no "programmatic" issues with that.

But the issue is : the servlet is registered too early when `ResourceProvider`  is not yet available.
Since the developer doesn't not anything about `ResourceProvider`  internal SPI it becomes an issue.

Please note: _this is not an issue of bundle deployment order_.
This is an issue with a servlet registration.

Once we implement an automatic servlet registration we will just force the waiting for the `ResourceProvider` service availability. 
But for now when developer has to register the servlet himself (and we will allow to do this anyway even with automatic servlet registration) we have to provide some mechanism which register a servlet only when there is already `ResourceProvider`. And I don't see how to do this properly.

`flow-osgi`  bundle is required for ANY Vaadin WAB. So it has to be deployed before the WAB.
Yes, it may be redeployed at any time but  the service should be available at the servlet registration only.
A Vaadin WAB may not be activated before its dependencies are activated. `flow-osgi`  is one of such dependencies.
So the activation of WAB should happen after the activation of `flow-osgi`.
So I think I may rely on the fact that a bundle activator of WAB is called (or any declarative service is registered) _after_ the `flow-osgi`  bundle Activator is called. 
So I think I can just move the service registration in the bundle Activator and it will just work.

Another option is to move the OSGi logic implemented in the https://github.com/vaadin/flow/blob/feature/osgi/flow-osgi/src/main/java/com/vaadin/flow/osgi/support/OSGiResourceProvider.java#L40 into `flow-server`  module but I don't like this way.

Reply:
I agree!
>> Please note: this is not an issue of bundle deployment order.
>> This is an issue with a servlet registration.

but most will try to fix this using the bundle deployment order. that would be wrong!

>> Another option is to move the OSGi logic implemented in the https://github.com/vaadin/flow/blob/feature/osgi/flow-osgi/src/main/java/com/vaadin/flow/osgi/support/OSGiResourceProvider.java#L40 into flow-server module but I don't like this way.

was my first idea. not sure that this is good.

>> Once we implement an automatic servlet registration we will just force the waiting for the ResourceProvider service availability.
But for now when developer has to register the servlet himself (and we will allow to do this anyway even with automatic servlet registration) we have to provide some mechanism which register a servlet only when there is already ResourceProvider. And I don't see how to do this properly.

I have to understand more why you use this ResourceProvider at the moment of activating the service

Reply:
This is an abstraction for strategy to access classpath resources.
In OSGi the resources can't be access the same way as in non-OSGi where resource is simply always in the classpath of the web app and you may access it via `SomeClass.class.getResource`.
In OSGi the correct way is `Bundle.getResource`  (and the Bundle here is not the same bundle as `flow-server`  bundle). So the resource access strategy is extracted to an abstraction layer and two implementations are provided:
for OSGi and non OSGi case.

Reply:
I would like to extract OSGi related code as much as possible to `flow-osgi` and don't have anything related to OSGi in `flow-server` or have a minimum things there.